### ISSUE TITLE
An attempt to simplify/clarify 2.4.11 Focus Appearance

### ISSUE THREAD

**Issue by yatil** [w3c/wcag#2687]
The [2.4.11 Focus Appearance Success Criterion](https://www.w3.org/TR/WCAG22/#focus-appearance) is long and hard to understand. The main obstacles are in my opinion the different clauses that allow for many permutations to meet this SC. This specificity and detail probably is more suitable to an AAA success criteria.

Generally, I applaud the Working Group for the effort of making clear what a “visible focus indicator” is. I have loosely followed the discussions and have noticed the effort of the Working Group participants working on this Success Criterion.

That said, coming at the SC with fresh eyes left a puzzled expression on my face, and various tweets and [blog posts (example)](https://hidde.blog/focus-appearance-too-complex/) have cemented it. The longer I sat with the SC, the more I think a simplification and clarification would be beneficial.

There are two goals of the success criterion: 
1. Contrast to adjacent pixels and pixels of the non-focused state. 
2. Defining the area that needs to be visible/change.

To allow for some simplification, I defined (pretty arbitrary) a minimum absolute focus area of 40px squared. That means if you have irregular shapes or circles or anything that is not defined by perimeters, you can fall back as a tester and see if 40px have changed. (For reference, I just took the smallest reasonable component as a 10×10 square) and multiplied the smallest side by 4. As I said, pretty arbitrary.)

I also tried to pull out repeating elements, like the 3:1 contrast ratio. Having these requirements repeat means that cognitively they appear to be two requirements, while there is only one value.

So, this is my proposal/idea with more rationale below:

> Keyboard focus indicators of components or subcomponents meet the following requirements when visible:
>
> 1. **Contrast**: A 3:1 contrast ratio is met
>    - between colors of pixels that change between the unfocused and the focused state of the component, and
>    - between the pixels of the focus indicator in the focused state and directly adjacent pixels.
> 2. **Area**: The minimum area of the focus indicator is either
>    - at least a 1 CSS pixel thick line enclosing the user interface component, or
>    - at least as large as a 4 CSS pixel thick line along the shortest side of the minimum bounding box of the unfocused component, or
>    - at least 40 CSS pixels squared.
> 
> Exceptions:
> - **Thick indicator**: If the focus indicator is at least a 2 CSS pixel thick line enclosing the user interface component, a minimum contrast ratio is not required between adjacent pixels of the indicator and the component.
> - **Not adjustable**: The focus indicator is determined by the user agent and cannot be adjusted by the author.
> - **Not modified**: The focus indicator and the indicator's background color are not modified by the author.

**Further Rationale**

> Keyboard focus indicators of components or subcomponents meet the following requirements when visible:

Trying to put as much repeating content into the opening as possible, especially regarding the components or subcomponents. (“When visible” comes from the current CR draft. As it is an AA criterion and _Focus Visible_ will be an A criterion, that feels redundant.)

> 1. **Contrast**: A 3:1 contrast ratio is met
>    - between colors of pixels that change between the unfocused and the focused state of the component, and
>    - between the pixels of the focus indicator in the focused state and directly adjacent pixels.

Contrast is currently split between two different ways of doing focus styles in the CR draft, but they both want to do the same. Making sure there is enough contrast between the pixels of the focus indicator in focused states to the pixels in the unfocused state, and between the focus indicator and pixels next to it. 

> 2. **Area**: The minimum area of the focus indicator is either
>    - at least a 1 CSS pixel thick line enclosing the user interface component, or
>    - at least as large as a 4 CSS pixel thick line along the shortest side of the minimum bounding box of the unfocused component, or
>    - at least 40 CSS pixels squared.

How to calculate the minimum area is one of the most complicated aspects of this SC in the CR draft. To simplify it, I thought of the options you should have: a border/outline around the component, a 4px thick line next to the component or some indicator of 40 CSS pixels squared. It prioritizes the technique that the Working Group wants to encourage as the first advice.

> Exceptions:
> - **Thick indicator**: If the focus indicator is at least a 2 CSS pixel thick line enclosing the user interface component, a minimum contrast ratio is not required between adjacent pixels of the indicator and the component.
> - **Not adjustable**: The focus indicator is determined by the user agent and cannot be adjusted by the author.
> - **Not modified**: The focus indicator and the indicator's background color are not modified by the author.

I left the exceptions as they are, apart from adding an exception that removes the requirement of contrast between the focus indicator and the adjacent pixels of the component in cases where the indicator is at least 2 CSS pixel thick. I am not sure if this is the right place, because it only exceptions part of the SC, but I didn’t feel it fit into the other categories. Possibly that sentence could be positioned after the list and before the exceptions.

**A few other notes:**

- I think this could be two SCs: “Focus Contrast” and “Focus Area”, and there are advantages and disadvantages to that.
- I wonder if it needs to be “40 CSS pixels squared”, but just saying “40 CSS pixels” did not feel correct for my “English as a second language” brain. If it doesn’t make a difference, I’d love to drop the “squared”.
- In general, most of the use cases for focus should be handled by browsers. Maybe, at some point in the future, authors could specify an “accent-color” to have some ownership of the focus styles (and browsers allow users to overwrite that with their own color preference). And of course if there are very specific needs, they would specify that they are in focus specification pro mode.
- I realize that this is a much less specific proposal than the one in the CR draft. This is on purpose in the attempt to make it easier to apply and test.
- Perhaps there are use cases that this idea does not catch, whereas the CR draft does. There is always the risk of that. Perhaps this SC is less ambitious or less specific than the CR draft. I personally think that it can be worth the tradeoff if the simplification leads to better usage.
- Sorry for this long post, but I hope showing the rationale helps to understand the proposal/idea.

👋 Eric

**Comment by WilcoFiers** [comment]
The big problem with this is that if the entire focus indicator needs a contrast of 3:1 with its unfocused state, focus indicators aren't allowed to fade out. It creates a situation where improving the focus indicator makes it fail.

Here's an example: https://codepen.io/wilcofiers/full/zYjPMLE

Link 1 has a simple 1px outline on focus. Link 2 also has that outline, but in addition, it gets a box-shadow that, because it fades results in a significant part of the focus indicator having a contrast below 3:1. Needing to support fades is a big reason of why the SC is as complicated as it is.

**Comment by yatil** [comment]
Hm. I see… I think that is addressable…

**Comment by brothercake** [comment]
> The big problem with this is that if the entire focus indicator needs a contrast of 3:1 with its unfocused state, focus indicators aren't allowed to fade out. It creates a situation where improving the focus indicator makes it fail.

As I understand it, this situation is covered by the Gradients section of the understanding docs. Only those parts of the focus indicator that have 3:1 contrast are considered to be part of the focus indicator for the purposes of evaluation. So a 1px solid outline with an additional box shadow would pass on the basis of the solid outline, the shadow part doesn't need to be considered.

**Comment by yatil** [comment]
Another try to be even more simple while still catching the different use cases. But I have to say, this is tough. I personally think it must change to be actionable and usable, but no variant that I have tried was sufficiently easy to understand.

> ### 2.4.11 Focus Contrast
> 
> A 3:1 contrast ratio is required when components receive focus:
> - for at least an area of pixels that changes from the unfocused state the size of 40 CSS pixels or a 1 CSS pixel thick line enclosing the component.
> - between these contrasting pixels and pixels adjacent to all pixels of the the focused component.
> - between these contrasting pixels and the component, unless the area of the contrasting pixels is at least a 2 CSS pixel thick line enclosing the component.
> 
> Exception: The focus indicator is determined by the user agent and not changed by the author.

**Comment by alastc** [comment]
> 40 CSS pixels

As a minimum bar, I think that's too low. It would mean the old Firefox default of a dotted line (that many people complained about) would pass for most controls (e.g. anything larger than 50px wide). 

It has to be proportional to the size of the component or you leave a gap (from the user point of view) for large components, or a difficult to pass requirement (for authors) for small components.

> But I have to say, this is tough.

Yep, we've been through the process of cutting it down, then accounting for the various use-cases several times.

**Comment by yatil** [comment]
> As a minimum bar, I think that's too low. 

I agree, but most bars in WCAG are fairly low, happy to come up with a higher threshold. Currently, 4 times the smallest side could easily be less than 40px, for example. Even 4 times 16 is only 64px, which I would also happily accept as the minimum. 

> then accounting for the various use-cases several times.

Yes, use case inflation is certainly an issue here. Maybe it would be easier to scope the SC just to links and buttons in the first step and then add other SCs for other UI elements? It’s pretty clear that it can’t go into the standard as it is, IMHO.

**Comment by alastc** [comment]
Suggested response for group consideration:

<hr>

Thank you for trying to simplify this SC. We have been through many iterations, hoping that someone could work out a way to simplify without undermining the aims.

There are several things the [proposal](https://github.com/w3c/wcag/issues/2687#issuecomment-1270722474) doesn't account for though, including:

- Keeping the indicator size proportional to the component, which is important for visibility and ease of implementation.
- "... pixels adjacent to all pixels of the the focused component", to match the same requirement, it can be all adjacent pixels. However, it is then confusing as to how gradients would be evaluated. ("These" pixels is less clear than "an area meets all the following").
- Given the need for some complexity to allow for intuitively visible indicators, the group thought that having two parts (a simple part and the more complex part) was helpful for understanding.

Some examples that highlight the need for proportionality have come up (which are both links/buttons) such as:

<img src="https://user-images.githubusercontent.com/216630/200836785-7aeba782-9564-4233-9c36-15bb1e227314.png" alt="Large card containing a picture, text, and visible link. There is a thin green line at the bottom." width="200">

With 40px, the thin green line at the bottom would be a passing indicator.

At the other end of the scale, this (from issue #1329) would struggle to have 40px:
<img src="https://user-images.githubusercontent.com/216630/200837444-db895738-f9b3-4d8c-942b-8c0100e00b0a.png" alt="two questions marks, the first is a blue underlined link, the second has a yellow background and thick black underline indicating the focus state.">

**Comment by alastc** [comment]
Response approved: https://www.w3.org/2022/11/22-ag-minutes#t11

**Comment by yatil** [comment]
FWIW, Example 1 is better than the current AA guidance of having a 1 by 1 pixel requirement. As for Example 2, the user could use the 4px width area of the smallest side as the basis for the focus style. Which would be much less than the 40px alternative.

The beauty of my initial proposal was that you could pick an implementation based on easy to verify parameters.

(Yes, this answer is 3,5 years late. My apologies. Standards burnout is real. No need to reply.)
